doc/internal/man7/EVP_PKEY.pod

```diff
@@ -8,37 +8,198 @@ EVP_PKEY - an internal description
-
-An B<EVP_PKEY> also contains a cache of provider side copies of the
-key, each adapted for the provider that is going to use that copy to
-perform some operation.
-For a legacy "origin", the B<EVP_PKEY_ASN1_METHOD>'s functions
-export_to() and dirty_cnt() must be implemented for such caching to be
-possible.  For a provider side "origin", the B<EVP_KEYMGMT>'s function
-OP_keymgmt_export() must be implemented.  In all cases, the receiving
-B<EVP_KEYMGMT> must have an implemented OP_keygmt_import().
+The private/public key pair that an B<EVP_PKEY> contains is refered to
+as its "internal key" or "origin" (the reason for "origin" is
+explained further down, in L</Export cache for provider operations>),
+and it can take one of the following forms:
+
+=over 4
+
+=item legacy origin
+
+This is the form that an B<EVP_PKEY> in OpenSSL prior to 3.0 had.  The
+internal key in the B<EVP_PKEY> is a pointer to to the low-level key
+types, such as B<RSA>, B<DSA> and B<EC>, or an engine driven
+structure, and is governed by an associated L<EVP_PKEY_METHOD(3)> and
+an L<EVP_PKEY_ASN1_METHOD(3)>.
+
+The functions available through those two method structures get full
+access to the B<EVP_PKEY> and therefore have a lot of freedom to
+modify whatever they want.  This also means that an B<EVP_PKEY> is a
+shared structure between libcrypto and any ENGINE that serves such
+methods.
+
+=item provider-native origin
+
+This is a new form in OpenSSL 3.0, which permits providers to hold the
+key data (see L<provider-keymgmt(7)>).  The internal key in the
+B<EVP_PKEY> is a pointer to that key data held by the provider, and
+is governed by an associated L<EVP_KEYMGMT(3)> method structure.
+
+The functions available through The L<EVP_KEYMGMT(3)> have no access
+to the B<EVP_PKEY>, and can therefore not make any direct changes.
+Similarly, the key data that the B<EVP_PKEY> points at is only known
+to the functions pointed at in the L<EVP_KEYMGMT(3)>.
+
+=back
+
+These two forms can never co-exist in the same B<EVP_PKEY>, the main
+reason being that having both at the same time will create problems
+with synchronising between the two forms, and potentially make it
+confusing which one of the two is the origin.
+
+=head2 Key mutability
+
+The B<EVP_PKEY> internal keys are mutable.
+
+This is especially visible with internal legacy keys, since they can
+be extracted with functions like L<EVP_PKEY_get0_RSA(3)> and then
+modified at will with functions like L<RSA_set0_key(3)>.
+
+Internal provider native keys are also possible to be modified, if the
+associated L<EVP_KEYMGMT(3)> implementation allows it.  This is done
+with L<EVP_PKEY_set_params(3)> and its specialised derivatives.  The
+OpenSSL providers allow it for the following:
+
+=over 4
+
+=item DH, EC, ED25519, ED448, X25519, X448:
+
+It's possible to set the encoded public key.  This is supported in
+particular through L<EVP_PKEY_set1_encoded_public_key(3)>.
+
+=item EC:
+
+It's possible to flip the ECDH cofactor mode.
```

Comment:
Why is this something we want to do?

Reply:
Backward compatibility.  OpenSSL 1.1.1 allows this:

``` C
EC_KEY_set_flags(key, EC_FLAG_COFACTOR_ECDH);
```

and

``` C
EC_KEY_clear_flags(key, EC_FLAG_COFACTOR_ECDH);
```

Personally, I deeply dislike this, but this is reported as something people want to do.  Or at least, @romen was worried that this "feature" would disappear..

Reply:
I am not deeply attached to it either, but it was just precaution as in 1.1.1 this flag can be set on the key object for its whole lifetime or just on the ctx for a single operation run.
With the PoV of "`EVP` must be feature complete and have a way to do all the things we could do with the lower-level functions that are being deprecated".

Flipping back and forth on the same key is likely a misuse of the functionality, but we did not enforce anything about it in 1.1.1.

Reply:
Ok, I think that resolves this conversation